File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week5/Queue.java
/*
Queue.java
Billy Cussen
13/11/2020
*/
public class Queue<T> implements QueueInterface<T>{

    Node<T> first;
    Node<T> last;

    public Queue(){
        this.first = null;
        this.last = null;
    }

    @Override
    public void enQueue(T element){
        Node<T> oldLast = this.last;
        this.last = new Node<T>(element, null);
        if(isEmpty()){
            this.first = this.last;
        } else{
            oldLast.next = this.last;
        }
    }

    
    @Override
    public T deQueue(){
        if(isEmpty()){
            throw new NullPointerException("Can't Dequeue, the Queue is empty!");
        } else{
            T element = first.element;
            first = first.next;
            return element;
        }
    }
    
    @Override
    public T peek(){
        if(isEmpty()){
            throw new NullPointerException("Can't peek, the Queue is Empty!");
        } else{
            return this.first.element;
        }
    }
    
    @Override
    public boolean isEmpty(){
        return this.first == null;
    }

    @Override
    public int size(){
        return sizeHelper(this.first);
    }

    private int sizeHelper(Node<T> current){
        if(current==null){
            return 0;
        }
        return 1 + sizeHelper(current.next);
    }


    @Override
    public String toString(){
        if(isEmpty()){
            throw new NullPointerException("Can't traverse, the Queue is Empty!");
        }
        String result = "***QUEUE TRAVERSAL***\n";
        int position = 0;
        return toStringHelper(this.first, position, result);
    }

    private String toStringHelper(Node<T> current, int position, String result){
        if(current == null){
            result+="***END***\n";
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        return toStringHelper(current.next,++position,result);
    }

}

File: Python/Data Structures & Algorithms Module - Python/Revision/Queues/Queue.py
"""
Queue.py
Billy Cussen
11/01/2021
"""
from Node import Node

class Queue:

    def __init__(self):
        self.first = None
        self.last = None

    def enQueue(self, element):
        oldLast = self.last
        self.last = Node(element, None)
        if self.isEmpty():
            self.first = self.last
        else:
            oldLast.next = self.last

    def deQueue(self):
        if self.isEmpty():
            raise Exception("Can't Dequeue, Queue is Empty!")
        else:
            elem = self.first.element
            self.first = self.first.next
            return elem

    def isEmpty(self):
        return self.first == None

    def size(self):
        if self.isEmpty():
            raise Exception("Can't get Size, Queue is Empty!")
        return self.sizeHelper(self.first)

    def sizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.sizeHelper(current.next)

    def toString(self):
        if self.isEmpty():
            return "Can't print Queue as it is Empty!"
        else:
            result = "***QUEUE TRAVERSAL***\n"
            position = self.size()-1
            return self.toStringHelper(self.first, result, position)

    def toStringHelper(self, current, result, position):
        if current == None:
            result+= "***END***\n"
            return result
        result += "Position "+str(position)+", Element: "+str(current.element)+"\n"
        position = position - 1
        return self.toStringHelper(current.next, result, position)

